rust/feature-flags/src/api/rate_limiter.rs

```diff
@@ -0,0 +1,364 @@
+use crate::api::{errors::FlagError, rate_parser::parse_rate_string};
+use common_metrics::inc;
+use common_types::TeamId;
+use governor::{clock, state::keyed::DefaultKeyedStateStore, Quota, RateLimiter};
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::num::NonZeroU32;
+use std::sync::{Arc, RwLock};
+use tracing::{info, warn};
+
+/// Type alias for a keyed rate limiter
+type KeyedRateLimiterInner<K> = Arc<RateLimiter<K, DefaultKeyedStateStore<K>, clock::DefaultClock>>;
+
+/// Type alias for the custom limiters map
+type CustomLimitersMap<K> = Arc<RwLock<HashMap<K, KeyedRateLimiterInner<K>>>>;
+
+/// Generic keyed rate limiter with default and per-key custom rate overrides
+///
+/// This limiter supports:
+/// - A configurable default rate limit for all keys
+/// - Custom per-key rate limits
+/// - Thread-safe concurrent access via Arc and RwLock
+/// - Configurable Prometheus metrics
+///
+/// # Type Parameters
+/// * `K` - The key type (e.g., TeamId, UserId, etc.)
+///
+/// # Algorithm
+/// Uses the GCRA (Generic Cell Rate Algorithm) via the `governor` crate,
+/// which is functionally equivalent to a leaky bucket but more efficient.
+#[derive(Clone)]
+pub struct KeyedRateLimiter<K>
+where
+    K: Hash + Eq + Clone + Display + Send + Sync + 'static,
+{
+    /// Default rate limiter for keys without custom rates
+    default_limiter: KeyedRateLimiterInner<K>,
+
+    /// Custom rate limiters for specific keys
+    /// Maps key → rate limiter
+    /// Wrapped in RwLock for thread-safe access
+    custom_limiters: CustomLimitersMap<K>,
+
+    /// Prometheus metric name for total requests
+    request_counter: &'static str,
+
+    /// Prometheus metric name for rate limited requests
+    limited_counter: &'static str,
+}
+
+/// Type alias for flag definitions rate limiting (per-team)
+pub type FlagDefinitionsRateLimiter = KeyedRateLimiter<TeamId>;
```

Comment:
I like this, and the implementation is solid, but is there a reason why you shipped your own rate limited instead of using, say, [tower_governor](https://github.com/benwis/tower-governor)? We definitely have more control over how we do rate limiting by shipping our own but it may mean that as we go, we'll need to keep contributing to this tool instead of using one that supports the features we need natively. Given that we're already importing `governor`, it seems like a natural extension to use this lib too.